orbit/models/module.py

```diff
@@ -0,0 +1,294 @@
+    """
+    def __init__(self, n_bootstrap_draws=-1, prediction_percentiles=None, **kwargs):
+        super().__init__(**kwargs)
+        self.n_bootstrap_draws = n_bootstrap_draws
+        self.prediction_percentiles = prediction_percentiles
+
+        # set default args
+        self._prediction_percentiles = None
+        self._n_bootstrap_draws = self.n_bootstrap_draws
+        self._set_default_args()
+
+    def _set_default_args(self):
+        if self.prediction_percentiles is None:
+            self._prediction_percentiles = [5, 95]
+        else:
+            self._prediction_percentiles = copy(self.prediction_percentiles)
+
+        self._prediction_percentiles += [50]  # always find median
+        self._prediction_percentiles = list(set(self._prediction_percentiles))  # unique set
+        self._prediction_percentiles.sort()
+
+        if not self.n_bootstrap_draws:
+            self._n_bootstrap_draws = -1
+
+    @staticmethod
+    def _bootstrap(num_samples, posterior_samples, n):
+        """Draw `n` number of bootstrap samples from the posterior_samples.
+
+        Args
+        ----
+        n : int
+            The number of bootstrap samples to draw
+
+        """
+        if n < 2:
+            raise IllegalArgument("Error: Number of bootstrap draws must be at least 2")
+
+        sample_idx = np.random.choice(range(num_samples), size=n, replace=True)
+        bootstrap_samples_dict = {}
+        for k, v in posterior_samples.items():
+            bootstrap_samples_dict[k] = v[sample_idx]
+
+        return bootstrap_samples_dict
+
+    def _full_bayes_predict(self, df, predict_func, **kwargs):
+        # raise if model is not fitted
+        if not self.is_fitted():
+            raise PredictionException("Model is not fitted yet.")
+        # if bootstrap draws, replace posterior samples with bootstrap
+        posterior_samples = self._bootstrap(
+            num_samples=self.estimator.num_sample,
+            posterior_samples=self._posterior_samples,
+            n=self._n_bootstrap_draws
+        ) if self._n_bootstrap_draws > 1 else self._posterior_samples
+
+        predicted_dict = predict_func(posterior_estimates=posterior_samples, df=df, include_error=True, **kwargs)
+        aggregated_df = aggregate_predictions(predicted_dict, self._prediction_percentiles)
+        aggregated_df = prepend_date_column(aggregated_df, df, self.date_col)
+        return aggregated_df
+
+
+class AggregatedPosteriorModule(BaseModule):
+    """ Abstract class for full aggregated posteriors prediction
+
+    In aggregated prediction, the parameter posterior samples are reduced using `aggregate_method`
+    before performing a single prediction.
+
+    Parameters
+    ----------
+    aggregate_method : { 'mean', 'median' }
+        Method used to reduce parameter posterior samples
+    """
+    def __init__(self, aggregate_method='mean', n_bootstrap_draws=5000, prediction_percentiles=None, **kwargs):
```

Comment:
here we use 5000 bootstrap draw; in MAPModule, we use 10000 bootstrap draw. do we want to make these two numbers the same?  A more important note is that why the default for AggregatedModule is not -1 (just like the full bayesian) where the mean or median will be calculated using all the samples instead of bootstrapped samples?